pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -548,9 +549,14 @@ private boolean processStreamEvents(MessageBatch messagesAndOffsets, long idlePi
       // this can be overridden by the decoder if there is a better indicator in the message payload
       RowMetadata msgMetadata = messagesAndOffsets.getMetadataAtIndex(index);
 
-      GenericRow decodedRow = _messageDecoder
-          .decode(messagesAndOffsets.getMessageAtIndex(index), messagesAndOffsets.getMessageOffsetAtIndex(index),
-              messagesAndOffsets.getMessageLengthAtIndex(index), reuse);
+      GenericRow decodedRow = null;
```

Comment:
agree, I think the idea is that the decoder encapsulates retry logic and error handling and only throw an exception  when it is exhausted to let the data manager retry ingestion. The decoder should return `null` if the payload cannot be decoded.

We have added documentation to `StreamMessageDecoder` however there are already OTB decoders which break this contract and users may still want to use them - this could also be the case of third-party plugins. We could make the decoders backwards incompatible, treat the issue as a bug and ignore this change instead but I think this may still be handy.

Reply:
I don't think the `StreamMessageDecoder` talks anything about retry / error handling. It only says that when return value is `null`, it means record is dropped. that's why I had this question. 

You bring up a good point about breaking contract with other message decoders. Perhaps in a follow-up PR, we should modify with a base decoder implementation that can handle exceptions and lets the OOTB catch-up. that way, we don't break anything, but provide a path to adapt for other implementations . 

I will resolve my comment for now. ty! :)